src/functions.php

```diff
@@ -343,9 +343,47 @@ function _checkTypehint(callable $callback, $object)
 
     $expectedException = $parameters[0];
 
-    if (!$expectedException->getClass()) {
+    // PHP before v8 used an easy API:
+    if (PHP_VERSION_ID < 80000) {
+        if (!$expectedException->getClass()) {
+            return true;
+        }
+
+        return $expectedException->getClass()->isInstance($object);
+    }
+
+    // Extract the type of the argument and handle different possibilities
+    $type = $expectedException->getType();
+    $types = [];
+
+    switch (true) {
+        case $type === null:
+            break;
+        case $type instanceof \ReflectionNamedType:
+            $types = [$type];
+            break;
+        case $type instanceof \ReflectionUnionType;
+            $types = $type->getTypes();
+            break;
```

Comment:
This will break again in PHP 8.1 with intersection types.

Reply:
Yes it will. But what do you suggest? Add a case for `\ReflectionUnionType` prematurely?

Reply:
As 8.1 is already in RC phase, it's not really prematurely, I guess.

Reply:
Well, I guess, you are right. Somehow this feels like it pushes the complexity over the "let's do this in this custom function" threshold. Especially if at a later point we get a mixture of intersections/unions this becomes a bit of a mess.

I just ran a quick test with `is_a`, hoping that at least one standard library function provides support for checking whether `$someObject` fulfills a (complex) type-contract. It seems though that this is not the case. Did I miss something?

Anyway, I see a few options:

1) Keep it as is, fix it when it comes up in the future. :trollface: 
2) Fix it exactly to spec, assuming that there is no composition/grouping of intersection and union-types
3) Build it recursively from the get-go to anticipate grouped/composited types

I guess 3) does not make sense, the API may even change until then. 1) would be an option if stdlib functionality is on the horizon. This makes me to opt for 2. What do you think?

Reply:
There are a couple of more options, e. g. throw for complex types, so not supporting them at all, or trying the invocation and catching the TypeError, but this might lead to different results if the TypeError isn't thrown from the callback itself.

Let's wait for the option of the maintainers on the best way forward.

Reply:
I kind of assumed (but didn't check) that you were also co-maintaining on the reactphp-side of things. Since this is not the case, let's wait for more feedback. 

In regards to your thoughts:
- Testing it via invocation feels a bit hacky
- Including checks that `::getTypes()` only returns `ReflectionNamedType` makes sense

Reply:
I added in a separate commit:

- basic support for intersection types
- an exception on complex/grouped types